File: server/services/face-detection/quality.ts
import type { DetectedFace } from "./detect";

/**
 * Compute a quality score (0-1) for a detected face.
 *
 * Factors:
 * - Face size relative to image (larger = better embedding)
 * - Detection confidence from SCRFD
 * - Landmark alignment quality (eyes level, nose centered, mouth symmetric)
 * - Bounding box aspect ratio (close to 1:1 is ideal)
 */
export function computeFaceQuality(
  face: DetectedFace,
  imageWidth: number,
  imageHeight: number,
): number {
  const sizeScore = computeSizeScore(face.box, imageWidth, imageHeight);
  const confidenceScore = face.confidence;
  const landmarkScore = computeLandmarkScore(face.landmarks, face.box);
  const aspectScore = computeAspectScore(face.box);

  // Weighted combination — confidence and size matter most
  return sizeScore * 0.3 + confidenceScore * 0.3 + landmarkScore * 0.25 + aspectScore * 0.15;
}

/**
 * Score based on face size relative to image area.
 * Faces covering more of the image produce better embeddings.
 * Uses a sigmoid-like curve: tiny faces score low, medium faces score well.
 */
export function computeSizeScore(
  box: DetectedFace["box"],
  imageWidth: number,
  imageHeight: number,
): number {
  const imageArea = imageWidth * imageHeight;
  if (imageArea <= 0) return 0;

  const faceArea = box.width * box.height;
  const ratio = faceArea / imageArea;

  // Sigmoid mapping: 0.001 -> ~0.1, 0.01 -> ~0.5, 0.05 -> ~0.85, 0.15+ -> ~1.0
  // k controls steepness, midpoint is the ratio that maps to 0.5
  const k = 120;
  const midpoint = 0.015;
  return 1 / (1 + Math.exp(-k * (ratio - midpoint)));
}

/**
 * Score based on landmark alignment quality.
 * Checks:
 * - Eyes roughly level (small vertical difference)
 * - Nose horizontally centered between eyes
 * - Mouth corners roughly symmetric relative to nose
 */
export function computeLandmarkScore(landmarks: number[][], box: DetectedFace["box"]): number {
  if (landmarks.length < 5) return 0.5;

  const [leftEye, rightEye, nose, leftMouth, rightMouth] = landmarks as [
    number[],
    number[],
    number[],
    number[],
    number[],
  ];

  const faceSize = Math.max(box.width, box.height);
  if (faceSize <= 0) return 0.5;

  // Eye level score: how horizontal are the eyes?
  const eyeVerticalDiff = Math.abs(leftEye[1]! - rightEye[1]!) / faceSize;
  const eyeLevelScore = Math.max(0, 1 - eyeVerticalDiff * 5);

  // Nose centering: is nose horizontally between the eyes?
  const eyeMidX = (leftEye[0]! + rightEye[0]!) / 2;
  const noseOffsetX = Math.abs(nose[0]! - eyeMidX) / faceSize;
  const noseCenterScore = Math.max(0, 1 - noseOffsetX * 4);

  // Mouth symmetry: are mouth corners roughly equidistant from nose?
  const leftMouthDist = Math.abs(leftMouth[0]! - nose[0]!);
  const rightMouthDist = Math.abs(rightMouth[0]! - nose[0]!);
  const maxMouthDist = Math.max(leftMouthDist, rightMouthDist, 1);
  const mouthSymmetry = 1 - Math.abs(leftMouthDist - rightMouthDist) / maxMouthDist;

  return eyeLevelScore * 0.4 + noseCenterScore * 0.35 + mouthSymmetry * 0.25;
}

/**
 * Score based on bounding box aspect ratio.
 * Faces are roughly square; extreme aspect ratios indicate partial/occluded faces.
 */
export function computeAspectScore(box: DetectedFace["box"]): number {
  if (box.width <= 0 || box.height <= 0) return 0;
  const ratio = box.width / box.height;
  // Ideal ratio is 1.0. Penalize deviation.
  // ratio 0.7-1.3 -> high score, <0.5 or >2.0 -> low score
  const deviation = Math.abs(ratio - 1.0);
  return Math.max(0, 1 - deviation * 1.5);
}

File: backend/internal/services/facedetection/quality.go
package facedetection

import "math"

// ComputeFaceQuality scores a detected face from 0.0 to 1.0 based on:
// - Size score (30%): sigmoid of faceArea / imageArea
// - Confidence score (30%): direct detection confidence
// - Landmark score (25%): eye levelness, nose centering, mouth symmetry
// - Aspect score (15%): penalize non-square bounding boxes
func ComputeFaceQuality(face DetectedFace, imgW, imgH int) float64 {
	sizeScore := computeSizeScore(face.Box, imgW, imgH)
	confidenceScore := face.Confidence
	landmarkScore := computeLandmarkScore(face)
	aspectScore := computeAspectScore(face.Box)

	return 0.30*sizeScore + 0.30*confidenceScore + 0.25*landmarkScore + 0.15*aspectScore
}

// computeSizeScore uses a sigmoid to map faceArea/imageArea to 0..1.
// Larger faces in the image score higher.
func computeSizeScore(box BoundingBox, imgW, imgH int) float64 {
	imageArea := float64(imgW) * float64(imgH)
	if imageArea == 0 {
		return 0
	}
	faceArea := box.Width * box.Height
	ratio := faceArea / imageArea

	// Sigmoid centered around 2% of image area, with moderate steepness
	return sigmoid((ratio - 0.02) * 200)
}

// computeLandmarkScore evaluates geometric quality from landmark positions:
// - Eye levelness: how horizontal the eye line is
// - Nose centering: nose tip centered between eyes
// - Mouth symmetry: mouth corners level and centered
func computeLandmarkScore(face DetectedFace) float64 {
	lm := face.Landmarks

	// Eye levelness: angle between eye line and horizontal
	leftEye := lm[0]
	rightEye := lm[1]
	eyeDist := math.Sqrt(sqDist(leftEye, rightEye))
	if eyeDist < 1 {
		return 0.5 // Can't compute meaningful score
	}

	eyeAngle := math.Abs(math.Atan2(rightEye[1]-leftEye[1], rightEye[0]-leftEye[0]))
	eyeLevelScore := math.Max(0, 1.0-eyeAngle*5) // Penalize rotation

	// Nose centering: nose should be roughly between eyes horizontally
	nose := lm[2]
	eyeMidX := (leftEye[0] + rightEye[0]) / 2
	noseOffsetX := math.Abs(nose[0]-eyeMidX) / eyeDist
	noseCenterScore := math.Max(0, 1.0-noseOffsetX*3)

	// Mouth symmetry: mouth corners should be level
	leftMouth := lm[3]
	rightMouth := lm[4]
	mouthDist := math.Sqrt(sqDist(leftMouth, rightMouth))
	if mouthDist < 1 {
		return (eyeLevelScore + noseCenterScore) / 2
	}
	mouthAngle := math.Abs(math.Atan2(rightMouth[1]-leftMouth[1], rightMouth[0]-leftMouth[0]))
	mouthSymScore := math.Max(0, 1.0-mouthAngle*5)

	return (eyeLevelScore + noseCenterScore + mouthSymScore) / 3
}

// computeAspectScore penalizes bounding boxes that deviate from square.
func computeAspectScore(box BoundingBox) float64 {
	if box.Width == 0 || box.Height == 0 {
		return 0
	}
	ratio := box.Width / box.Height
	if ratio > 1 {
		ratio = 1 / ratio
	}
	// ratio is now in (0, 1], with 1 being perfectly square
	return ratio
}

func sigmoid(x float64) float64 {
	return 1.0 / (1.0 + math.Exp(-x))
}

func sqDist(a, b [2]float64) float64 {
	dx := a[0] - b[0]
	dy := a[1] - b[1]
	return dx*dx + dy*dy
}
